docs/k2htpdtorsvr.1
.TH k2htpdtorsvr "1" "February 2015" "k2htpdtorsvr" "K2HASH TRANSACTION"
.SH NAME
k2htpdtorsvr \- receiver server process for K2HASH TRANSACTION PLUGIN k2htpdtor.
.SH SYNOPSIS
.TP
\fB2htpdtorsvr
.SH DESCRIPTION
.TP 2
k2htpdtorsvr is a receiver for k2htpdtor K2HASH TRANSACTION PLUGIN. This program is expanding the receive data from k2htpdtor plugin through chmpx. This makes local k2hash file and writes/deletes keys/subkeys/values which is received from chmpx server side process.
.PP
.SH OPTIONS
.TP
\fB\-h\fR
display help
.TP
\fB\-ver\fR
display version
.TP
\fB\-conf\fR [file path]
specify configuration file(.ini .yaml .json).
.TP
\fB\-json\fR [file path]
specify configuration json string.
.TP
\fB\-d\fR [ERR | WAN | INFO | DUMP]
specify debugging level, chmpxbench displays debugging message for itself.
.TP
\fB\-dl\fR
specify sending message data length by byte. default value and minimum value are 64 bytes.
.br
.SH CONFIGURATION
.TP 0
The configuration is formatted "INI" or YAML or JSON file, and you can use JSON string.
.TP 2
Following keys in configuration:
.IP K2HTPDTORSVR([K2HTPDTORSVR]) 20
This program configuration section.
.br
.BR
.br
.IP OUTPUTFILE 20
output file path
.IP PLUGIN 20
plugin program which is called by this program at each receiving transaction data.
.IP K2HTYPE 20
k2hash type when using k2hash. you can specify MEMORY(MEM, M) or FILE(F) or TEMPORARY(TEMP, T).
.IP K2HFILE 20
k2hash file path when using k2hash file type.
.IP K2HFULLMAP 20
full mapping mode for k2hash mode when using k2hash
.IP K2HINIT 20
flag for initializing k2hash at starting when using k2hash
.IP K2HMASKBIT 20
initial mask bit count for k2hash when using k2hash
.IP K2HCMASKBIT 20
initial cmask bit count for k2hash when using k2hash
.IP K2HMAXELE 20
initial element count for k2hash when using k2hash
.IP K2HPAGESIZE 20
initial page size for k2hash when using k2hash
.IP K2HDTORSVR_TRANS 20
if relay transaction, specify transaction chmpx configuration.
.IP DTORTHREADCNT 20
if relay transaction, specify transaction thread count to setting k2hash.
.br
.SH ENVIRONMENT
.TP 2
This program use some environments automatically, please take care these environments.
.IP DTORSVRCONFFILE 20
specify configuration file by environment, this program uses this value when configuration file is not specified in command line.
.IP DTORSVRJSONCONF 20
specify configuration json string by environment, chmpx uses this value when configuration json string is not specified in command line.
.SH NOTES
.TP
K2HASH is NoSQL(key value store) library and tools for this library by Yahoo! JAPAN. K2HASH is following specifications.
.SH SEE ALSO
.TP
k2hash(1) chmpx(1) fullock(3) k2htpdtor(3)
.SH BUGS
.TP
Please report owner and committer.
.SH AUTHOR
k2htpdtorsvr has been written by Takeshi Nakatani.


ChangeLog
k2htpdtor (1.0.28) trusty; urgency=low

  * Fixed misspelling, and etc - #18
  * Reflected the spelling check result for all sources(2) - #17
  * Added check using cppcheck - #16
  * Reflected the spelling check result for all sources - #15

 -- Takeshi Nakatani <ggtakec@gmail.com>  Thu, 17 Jan 2019 11:04:06 +0900

k2htpdtor (1.0.27) trusty; urgency=low

  * Updated common build tool files

 -- Takeshi Nakatani <ggtakec@gmail.com>  Tue, 11 Dec 2018 14:03:07 +0900

k2htpdtor (1.0.26) trusty; urgency=low

  * Changed to strict packaging and changed documents

 -- Takeshi Nakatani <ggtakec@gmail.com>  Thu, 06 Dec 2018 15:36:40 +0900

k2htpdtor (1.0.25) trusty; urgency=low

  * Fixed destructing code for corresponding to SIOF, etc

 -- Takeshi Nakatani <ggtakec@gmail.com>  Wed, 17 Oct 2018 19:12:19 +0900

k2htpdtor (1.0.24) trusty; urgency=low

  * Revert subdir-objects automake option in src directory
  * avoid static object initialization order problem(SIOF)

 -- Takeshi Nakatani <ggtakec@gmail.com>  Mon, 02 Jul 2018 11:11:30 +0900

k2htpdtor (1.0.23) trusty; urgency=low

  * Cleanup credit comments in sources and Modified etc
  * Changed build logic to new for rpm/debian packaging
  * Fixed a filename typo of make_variables.sh

 -- Takeshi Nakatani <ggtakec@gmail.com>  Thu, 17 May 2018 14:21:10 +0900

k2htpdtor (1.0.22) trusty; urgency=low

  * Updated submodule dependency for chmpx/k2hash

 -- Hirotaka Wakabayashi <hiwakaba@yahoo-corp.jp>  Mon, 26 Mar 2018 15:44:29 +0900

k2htpdtor (1.0.21) trusty; urgency=low

  * Added lib_version_for_link option to make_valiables.sh
  * Updated submodule dependency for chmpx/k2hash/fullock

 -- Takeshi Nakatani <ggtakec@gmail.com>  Tue, 28 Mar 2017 13:20:11 +0900

k2htpdtor (1.0.20) trusty; urgency=low

  * Updated dependency for fixing bugs in libfullock
  * Updated submodules and changed .travis.yaml - #1

 -- Takeshi Nakatani <ggtakec@gmail.com>  Tue, 14 Feb 2017 16:54:17 +0900

k2htpdtor (1.0.19) trusty; urgency=low

  * First version of open source on Github

 -- Takeshi Nakatani <ggtakec@gmail.com>  Wed, 07 Dec 2016 19:10:45 +0900



src/k2htpdtorsvr.cc
/*
 * k2htpdtor for K2HASH TRANSACTION PLUGIN.
 *
 * Copyright 2015 Yahoo Japan Corporation.
 *
 * K2HASH TRANSACTION PLUGIN is programmable I/F for processing
 * transaction data from modifying K2HASH data.
 *
 * For the full copyright and license information, please view
 * the license file that was distributed with this source code.
 *
 * AUTHOR:   Takeshi Nakatani
 * CREATE:   Tue Mar 3 2015
 * REVISION:
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>

#include <k2hash/k2hash.h>
#include <k2hash/k2hutil.h>
#include <k2hash/k2hdbg.h>
#include <chmpx/chmpx.h>

#include <string>
#include <map>

#include "k2htpdtorsvrman.h"
#include "k2htpdtorsvrparser.h"

using namespace std;

//---------------------------------------------------------
// Types
//---------------------------------------------------------
typedef std::map<std::string, std::string>	dtorparam_t;

//---------------------------------------------------------
// Variable
//---------------------------------------------------------
static volatile bool	is_doing_loop = true;

//---------------------------------------------------------
// Version
//---------------------------------------------------------
extern char k2htpdtorsvr_commit_hash[];

//---------------------------------------------------------
// Functions
//---------------------------------------------------------
// [Usage]
// -h                help display
// -ver              version display
// -conf <ini file>  configuration file for k2htpdtorsvr processing
// -d <debug level>  debugging level: ERR(default) / WAN / INFO / SILENT
// -dlog <file path> output file for debugging message(default stderr)
// 
static void PrintUsage(char* argv)
{
	string	prgname = argv ? basename(argv) : "k2htpdtorserver";

	cout << "[Usage]" << endl;
	cout << prgname << " -h" << endl;
	cout << prgname << " [-conf <path> | -json <string>] [-d <debug level>] [-dlog <file path>]" << endl;
	cout << endl;
	cout << "-h                     help display" << endl;
	cout << "-ver                   version display" << endl;
	cout << "-conf <path>           configuration file(.ini .yaml .json) path" << endl;
	cout << "-json <json string>    configuration json string" << endl;
	cout << "-d <debug level>       debugging level: ERR(default) / WAN / INFO(MSG) / SILENT" << endl;
	cout << "-dlog <file path>      output file for debugging message(default stderr)" << endl;
	cout << endl;
}

static void PrintVersion(FILE* stream)
{
	static const char format[] =
		"\n"
		"k2htpdtorsvr Version %s (commit: %s)\n"
		"\n"
		"Copyright(C) 2015 Yahoo Japan Corporation.\n"
		"\n"
		"K2HASH TRANSACTION PLUGIN is programmable I/F for processing\n"
		"transaction data from modifying K2HASH data.\n"
		"\n";

	if(!stream){
		stream = stdout;
	}
	fprintf(stream, format, VERSION, k2htpdtorsvr_commit_hash);

	chmpx_print_version(stream);
	k2h_print_version(stream);
}

static bool ParamParser(int argc, char** argv, dtorparam_t& params)
{
	params.clear();

	bool	is_conf_opt	= false;
	bool	is_json_opt	= false;
	for(int cnt = 1; cnt < argc; cnt++){
		if(0 == strcasecmp(argv[cnt], "-h") || 0 == strcasecmp(argv[cnt], "-help")){
			params["h"] = "";

		}else if(0 == strcasecmp(argv[cnt], "-ver") || 0 == strcasecmp(argv[cnt], "-version")){
			params["v"] = "";

		}else if(0 == strcasecmp(argv[cnt], "-conf")){
			if(argc <= cnt + 1){
				fprintf(stderr, "[ERROR] Option %s needs parameter.\n", argv[cnt]);
				return false;
			}
			if(!is_file_safe_exist(argv[cnt + 1])){
				fprintf(stderr, "[ERROR] %s file does not regular file.\n", argv[cnt]);
				return false;
			}
			if(is_json_opt){
				fprintf(stderr, "[ERROR] Already set -json option, option %s could not be specified with -json.\n", argv[cnt]);
				return false;
			}
			is_conf_opt		= false;
			params["conf"]	= argv[++cnt];

		}else if(0 == strcasecmp(argv[cnt], "-json")){
			if(argc <= cnt + 1){
				fprintf(stderr, "[ERROR] Option %s needs parameter.\n", argv[cnt]);
				return false;
			}
			if(is_conf_opt){
				fprintf(stderr, "[ERROR] Already set -conf option, option %s could not be specified with -conf.\n", argv[cnt]);
				return false;
			}
			is_json_opt		= false;
			params["json"]	= argv[++cnt];

		}else if(0 == strcasecmp(argv[cnt], "-d")){
			if(argc <= cnt + 1){
				fprintf(stderr, "[ERROR] Option %s needs parameter.\n", argv[cnt]);
				return false;
			}
			if(0 != strcasecmp(argv[cnt + 1], "silent") && 0 != strcasecmp(argv[cnt + 1], "err") && 0 != strcasecmp(argv[cnt + 1], "wan") && 0 != strcasecmp(argv[cnt + 1], "info") && 0 != strcasecmp(argv[cnt + 1], "msg")){
				fprintf(stderr, "[ERROR] Option %s value %s is not defined.\n", argv[cnt], argv[cnt + 1]);
				return false;
			}
			params["d"] = lower(argv[++cnt]);

		}else if(0 == strcasecmp(argv[cnt], "-dlog")){
			if(argc <= cnt + 1){
				fprintf(stderr, "[ERROR] Option %s needs parameter.\n", argv[cnt]);
				return false;
			}
			params["dlog"] = argv[++cnt];

		}else{
			fprintf(stderr, "[ERROR] Unknown option %s.\n", argv[cnt]);
			return false;
		}
	}
	return true;
}

//---------------------------------------------------------
// Main
//---------------------------------------------------------
int main(int argc, char** argv)
{
	dtorparam_t	params;

	// Parse parameter
	// cppcheck-suppress unmatchedSuppression
	// cppcheck-suppress stlSize
	if(!ParamParser(argc, argv, params) || 0 == params.size()){
		PrintUsage(argv[0]);
		exit(EXIT_FAILURE);
	}

	// help
	if(params.end() != params.find("h")){
		PrintUsage(argv[0]);
		exit(EXIT_SUCCESS);
	}
	// version
	if(params.end() != params.find("v")){
		PrintVersion(NULL);
		exit(EXIT_SUCCESS);
	}

	// debug mode / log
	if(params.end() == params.find("d")){
		k2h_set_debug_level_silent();
		chmpx_set_debug_level_silent();
	}else if(params["d"] == "err"){
		k2h_set_debug_level_error();
		chmpx_set_debug_level_error();
	}else if(params["d"] == "wan"){
		k2h_set_debug_level_warning();
		chmpx_set_debug_level_warning();
	}else if(params["d"] == "info" || params["d"] == "msg"){
		k2h_set_debug_level_message();
		chmpx_set_debug_level_message();
	}else{	// params["d"] == "silent"
		k2h_set_debug_level_silent();
		chmpx_set_debug_level_silent();
	}
	if(params.end() == params.find("dlog")){
		if(!k2h_set_debug_file(params["dlog"].c_str()) || !chmpx_set_debug_file(params["dlog"].c_str())){
			ERR_K2HPRN("Could not set debug log file(%s).", params["dlog"].c_str());
			exit(EXIT_FAILURE);
		}
	}

	// get configuration option
	string	config;
	if(params.end() != params.find("conf")){
		config = params["conf"].c_str();
	}else if(params.end() != params.find("json")){
		config = params["json"].c_str();
	}

	// make and check info from conf file
	K2HTPDTORSVRINFO	info;
	if(!ParseK2htpdtorsvrConfiguration(config.c_str(), &info)){
		ERR_K2HPRN("Something wrong in conf file(%s).", config.c_str());
		exit(EXIT_FAILURE);
	}

	// manager
	K2HtpSvrManager	SvrMan;
	if(!SvrMan.Initialize(&info)){
		ERR_K2HPRN("Failed to initialize manager object.");
		exit(EXIT_FAILURE);
	}

	// do processing
	if(!SvrMan.Processing()){
		ERR_K2HPRN("Something wrong occurred for loop.");
		SvrMan.Clean();
		exit(EXIT_FAILURE);
	}
	SvrMan.Clean();

	exit(EXIT_SUCCESS);
}

/*
 * VIM modelines
 *
 * vim:set ts=4 fenc=utf-8:
 */
